enhancements/authentication/oauth-resource-storage.md

```diff
@@ -0,0 +1,213 @@
+replaces:
+superseded-by:
+---
+
+# Secure OAuth Resource Storage
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. Do we really need token read access for users. This enhancement allows us to add that with reasonable effort,
+   but we should consider also either some imperative logout subresource, or an endpoint for the oauth server. This
+   will be simpler to implement, but lack the `oc get oauthaccesstokens` ability for normal users.   
+
+## Summary
+
+OAuthAccessTokens and OAuthAuthorizeToken use the object name for the security 
+sensitive token information. Hence, encrypting the value only in etcd means that 
+the token itself is stored as plain text in the etcd database and in encrypted
+etcd backups, i.e. unprotected.
+
+This enhancement is about the migration to a different storage format where the
+token object name is insensitive and can be therefore stored in plain text without
+risk.
+
+While changing the storage format, we will prepare for different token formats in
+the future by introducing a prefix syntax. The prefix will be used by the 
+authenticator to distinguish between the old and new verification semantics.
+
+## Motivation
+
+Etcd encryption without encrypted oauth access and authorize token is of limited
+value. While these tokens expire after a day usually (but this can be and is often
+customized by administrators to much longer times), the token from a fresh backup,
+the output of etcdctl or even stack traces or trace information in the apiserver
+ logs can be stolen and used to act as oauth authenticated user in the cluster.
+
+A side-effect of non-sensitive object names is that these can be exposed via the
+API to other parties than the owner of the token in the session using the token. 
+This is helpful to create an API that allows a user to list his token objects, 
+e.g. to implement a feature to log out all "other" tokens in a REST-ful way.
+
+### Goals
+
+- Change the `oauthaccesstokens` and `oauthauthorizetokens` storage format and 
+  schema such that the etcd keys alone (which are not encrypted when encryption 
+  at REST is enabled) are not sensitive.
+- Make a "log out all other sessions" feature possible in the future.
+
+### Non-Goals
+
+- Design the "log out all other sessions" feature.
+
+## Proposal
+
+We prefix oauth access and oauth authorize token names with a prefix:
+
+- no prefix for old tokens,
+- `"sha256:"` for new tokens. The name of the token objects is the sha256 hash 
+  of the actual token (not decoded, but plain base64) encoded as base64. The bearer
+  token will also carry the prefix, but have the token in plain text.
+
+In v4.(n-1) and in v4.n:
```

Comment:
My only concern would be what N and (N-1) are if a user attempts and emergency unsupported rollback to N-1. And if we are certain that their N-1 has support for `sha256:` tokens.

Reply:
Worst that can happen is that users have relogin.

Reply:
To be clear: this means that we add sha256 support in 4.n-1.x. If the customer does a downgrade to that, everything keeps working. If the the customer does a downgrade to 4.n-1.x-1, then users have to relogin, but otherwise things keep working. So it is similar as tokens that expire (after a day by default), but enforced by the unsupported downgrade.